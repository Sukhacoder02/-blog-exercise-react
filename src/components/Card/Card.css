.card {
  height: 20%;
  width: 30%;
  /* background-color: aqua; */
  border: 1px solid lightgrey;
}
.card .img {
  height: 60%;
}
.card .belowImage {
  height: 40%;
}
.card .img img {
  width: 100%;
  height: 100%;
  object-fit: cover;
}
.belowImage {
  padding: 5px 5%;
}
.textBelowImage {
  width: 100%;
  height: 100%;
  display: flex;
  justify-content: space-between;
  font-size: 10px;
  color: grey;
  padding-bottom: 2%;
}
.title {
  padding-top: 5px;
  padding-bottom: 5px;
}
.description {
  color: grey;
  font-size: 12px;
  padding-bottom: 5px;
}
hr {
  border: none;
  border-top: 1px solid lightgrey;
}
.reactions {
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding-top: 5px;
  /* padding-bottom: 5px; */
}
.hands {
  display: flex;
}
.hands p {
  color: grey;
  padding-left: 2px;
}
.reactions img {
  width: 20px;
  height: 20px;
}
.hands button {
  background-color: white;
  border: none;
  outline: none;
  cursor: pointer;
}
.heart button {
  background-color: white;
  border: none;
  outline: none;
  cursor: pointer;
}
