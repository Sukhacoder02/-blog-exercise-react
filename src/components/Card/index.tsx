import * as React from 'react';
import './Card.css';

interface CardData {
  date: string;
  readingTime: string;
  title: string;
  description: string;
  claps: number;
  image: string;
  liked: boolean;
}

const Card = (cardData: CardData): JSX.Element => {
  const [claps, setClaps] = React.useState(cardData.claps);

  const blackHeart = 'heart-black.svg';
  const redHeart = 'heart-red.svg';
  const [_redHeart, setHeart] = React.useState(cardData.liked ? true : false);
  const [disabled, setDisabled] = React.useState(false);

  return (
    <div className="card">
      <div className="img">
        <img src={process.env.PUBLIC_URL + `/Images/${cardData.image}`} alt="Image" className="cardImage" />
      </div>
      <div className="belowImage">
        <div className="textBelowImage">
          <p>{cardData.date}</p>
          <p>{cardData.readingTime}</p>
        </div>
        <div className="title">
          <h3>{cardData.title}</h3>
        </div>
        <div className="description">
          <p>{cardData.description}</p>
        </div>
        <hr />
        <div className="reactions">
          <div className="hands">
            <button
              data-testid="testClapButton"
              className="claps"
              disabled={disabled}
              onClick={() => {
                setClaps(claps + 1);
                setDisabled(true);
              }}>
              <img src={process.env.PUBLIC_URL + '/Icons/clapping.svg'} alt="image" />
            </button>
            <p data-testid="claps">{claps}</p>
          </div>
          <div className="heart">
            <button
              data-testid="testHeartButton"
              onClick={() => {
                setHeart(!_redHeart);
              }}>
              <img src={process.env.PUBLIC_URL + '/Icons/' + `${_redHeart ? redHeart : blackHeart}`} alt="image" />
            </button>
          </div>
        </div>
      </div>
    </div>
  );
};

export default Card;
