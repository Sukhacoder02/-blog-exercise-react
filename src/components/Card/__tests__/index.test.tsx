import * as React from 'react';
import Card from '../../Card';
import { fireEvent, render } from '@testing-library/react';

const mockProps = {
  date: '2nd January, 2018',
  readingTime: '2 mins',
  title: 'The future of abstract art and the culture ...',
  description:
    'Create a blog post subtitle that summarizes your post in a few short, punchy sentences and entices your...',
  claps: 10,
  liked: false,
  image: 'abstract.png',
};
const card = (
  <Card
    key={0}
    date={mockProps.date}
    readingTime={mockProps.readingTime}
    title={mockProps.title}
    description={mockProps.description}
    claps={mockProps.claps}
    image={mockProps.image}
    liked={mockProps.liked}
  />
);

describe('Component Card', () => {
  it('Shoud render without crashing', () => {
    const { asFragment } = render(card);
    expect(asFragment()).toMatchSnapshot();
  });
  describe('Button claps', () => {
    it('Should  increase the number of claps by 1 when pressed', () => {
      const { getByTestId } = render(card);
      const numClaps = getByTestId('claps');
      expect(numClaps.textContent).toBe(mockProps.claps.toString());
      const clapsButton = getByTestId('testClapButton');
      fireEvent.click(clapsButton);
      expect(numClaps.textContent).toBe((mockProps.claps + 1).toString());
    });
    it('Should be disabled after being pressed', () => {
      const { getByTestId } = render(card);
      const clapsButton = getByTestId('testClapButton') as HTMLButtonElement;
      fireEvent.click(clapsButton);
      expect(clapsButton.disabled).toBe(true);
    });
  });
  describe('Button like', () => {
    it('Should fill the button with red color when clicked', () => {
      const { getByTestId } = render(card);
      const heartButton = getByTestId('testHeartButton');
      expect(heartButton.querySelector('img')?.src.includes('black')).toBe(true);
      fireEvent.click(heartButton);
      expect(heartButton.querySelector('img')?.src.includes('red')).toBe(true);
    });
    it('Should fill the button with black color when clicked twice', () => {
      const { getByTestId } = render(card);
      const heartButton = getByTestId('testHeartButton');
      expect(heartButton.querySelector('img')?.attributes.src.value.includes('red')).toBe(false);
      fireEvent.click(heartButton);
      expect(heartButton.querySelector('img')?.attributes.src.value.includes('red')).toBe(true);
      fireEvent.click(heartButton);
      expect(heartButton.querySelector('img')?.attributes.src.value.includes('black')).toBe(true);
    });
  });
});
