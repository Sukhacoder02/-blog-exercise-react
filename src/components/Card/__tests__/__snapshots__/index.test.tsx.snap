// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Card Shoud render without crashing 1`] = `
<DocumentFragment>
  <div
    class="card"
  >
    <div
      class="img"
    >
      <img
        alt="Image"
        class="cardImage"
        src="/Images/abstract.png"
      />
    </div>
    <div
      class="belowImage"
    >
      <div
        class="textBelowImage"
      >
        <p>
          2nd January, 2018
        </p>
        <p>
          2 mins
        </p>
      </div>
      <div
        class="title"
      >
        <h3>
          The future of abstract art and the culture ...
        </h3>
      </div>
      <div
        class="description"
      >
        <p>
          Create a blog post subtitle that summarizes your post in a few short, punchy sentences and entices your...
        </p>
      </div>
      <hr />
      <div
        class="reactions"
      >
        <div
          class="hands"
        >
          <button
            class="claps"
            data-testid="testClapButton"
          >
            <img
              alt="image"
              src="/Icons/clapping.svg"
            />
          </button>
          <p
            data-testid="claps"
          >
            10
          </p>
        </div>
        <div
          class="heart"
        >
          <button
            data-testid="testHeartButton"
          >
            <img
              alt="image"
              src="/Icons/heart-black.svg"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
